modAionImpl/src/org/aion/zero/impl/config/dynamic/DynamicConfigKeyRegistry.java

```diff
@@ -0,0 +1,100 @@
+ *
+ * It is important to ensure that the getter returns a type that has a properly-defined equals
+ * method.  Apart from this, there is no restriction, but it is generally a good idea for them
+ * to return "atomic" types (i.e. primitives or their object equivalents).
+ */
+public class DynamicConfigKeyRegistry {
+    private final
+    Map<String, Pair<Function<Cfg,?>, IDynamicConfigApplier>> key2GetterApplier; // config_key -> (getter,applier)
+
+    public DynamicConfigKeyRegistry() {
+        this(DEFAULT_GETTERS_APPLIERS);
+    }
+
+    @VisibleForTesting
+    DynamicConfigKeyRegistry(Map<String, Pair<Function<Cfg,?>, IDynamicConfigApplier>> key2GetterApplier) {
+        this.key2GetterApplier = key2GetterApplier; // note: not a deep copy
+    }
+
+    public void bind(String key, Function<Cfg,?> getter, IDynamicConfigApplier applier) {
+        key2GetterApplier.put(key, new ImmutablePair<>(getter, applier));
+    }
+
+    public Set<String> getBoundKeys() {
+        return key2GetterApplier.keySet();
+    }
+
+    public Function<Cfg,?> getGetter(String key) {
+        return key2GetterApplier.get(key).getLeft();
+    }
+
+    public IDynamicConfigApplier getApplier(String key) {
+        return key2GetterApplier.get(key).getRight();
+    }
+
+    private static Map<String, Pair<Function<Cfg,?>, IDynamicConfigApplier>>
+    DEFAULT_GETTERS_APPLIERS = new HashMap<>() {{
+        put("aion.consensus", ImmutablePair.of(cfg -> cfg.getConsensus(), NotImplementedThrower.INST));
+        // When MiningApplier changes reviewed, delete above line and uncomment below line.
+//        put("aion.consensus.mining", ImmutablePair.of(
+//                cfg -> ((CfgConsensusPow)cfg.getConsensus()).getMining(),
+//                new MiningApplier()));
+
+        // Below are sections where no config keys within are dynamically changeable
+        put("aion.api", ImmutablePair.of(cfg -> cfg.getApi(), NotImplementedThrower.INST));
+        put("aion.net.id", ImmutablePair.of(cfg -> cfg.getNet().getId(), NotImplementedThrower.INST));
+        put("aion.net.p2p", ImmutablePair.of(cfg -> cfg.getNet().getP2p(), NotImplementedThrower.INST));
+        put("aion.sync", ImmutablePair.of(cfg -> cfg.getSync(), NotImplementedThrower.INST));
+        put("aion.log", ImmutablePair.of(cfg -> cfg.getLog(), NotImplementedThrower.INST));
+        put("aion.db", ImmutablePair.of(cfg -> cfg.getDb(), NotImplementedThrower.INST));
+    }};
+
+    /**
+     * An applyer that just throws an exception.  Used for sections of the config that don't have
+     * implementation for dynamic config changes yet.
+     */
+    private static class NotImplementedThrower implements IDynamicConfigApplier {
+        public static final NotImplementedThrower INST = new NotImplementedThrower();
+
+        @Override
+        public InFlightConfigChangeResult apply(Cfg oldCfg, Cfg newCfg) throws InFlightConfigChangeException {
+            throw new InFlightConfigChangeNotAllowedException(
```

Comment:
The last time I was on a "should I use exceptions here" debate, I was on the "don't use exceptions" side... I also considered myself a C++ programmer at that time.  I wonder if there is any correlation between those two things :P 

The classic rule is thumb is "don't use exceptions for ordinary flow control", so it just boils down to what one considers "ordinary"...  I'm kind of on the fence about this one.  The class is called DynamicConfigKeyRegistry and it gets used by a class called InFlightConfigReceiver -- so trying to change a config dynamically when it is not possible seems like an exception to me.  On the other hand, I could easily use Optional.empty() instead of NotImplementedThrower and I think that is actually clearer.  

I think what makes the most sense is that Optional.empty() (instead of NotImplementedThrower) is used to signify that a config key can't be changed.  But then the code that uses this class (InFlightConfigReceiver line 101) checks if it got an Optional.empty() and if so, throw InFlightConfigChangeNotAllowedException.  

Let me know if you see any problems with that approach.